Source/WebCore/rendering/AttachmentLayout.mm

```diff
@@ -0,0 +1,397 @@
+        if (thumbnailIcon || icon) {
+            iconRect = FloatRect(FloatPoint((attachmentRect.width() / 2) - (iconSize.width() / 2), 0), iconSize);
+            yOffset += iconRect.height() + attachmentItemMargin;
+        }
+    } else {
+        NSDictionary *textAttributesTitle = @{
+            (__bridge id)kCTFontAttributeName: (__bridge id)attachmentTitleFont().get(),
+            (__bridge id)kCTForegroundColorAttributeName: (__bridge id)attachmentActionColor(attachment)
+        };
+
+        buildWrappedLines(action, attachmentActionFont().get(), textAttributesTitle, attachmentWrappingTextMaximumLineCount);
+    }
+
+    bool forceSingleLineTitle = !action.isEmpty() || !subtitle.isEmpty() || hasProgress;
+    NSDictionary *textAttributesTitle = @{
+        (__bridge id)kCTFontAttributeName: (__bridge id)attachmentTitleFont().get(),
+        (__bridge id)kCTForegroundColorAttributeName: (__bridge id)attachmentTitleColor(attachment)
+    };
+
+    buildWrappedLines(title, attachmentTitleFont().get(), textAttributesTitle, forceSingleLineTitle ? 1 : attachmentWrappingTextMaximumLineCount);
+    NSDictionary *textAttributesSubTitle = @{
+        (__bridge id)kCTFontAttributeName: (__bridge id)attachmentSubtitleFont().get(),
+        (__bridge id)kCTForegroundColorAttributeName: (__bridge id)attachmentSubtitleColor(attachment)
+    };
+    buildSingleLine(subtitle, attachmentSubtitleFont().get(), textAttributesSubTitle);
+
+    if (!lines.isEmpty()) {
+        for (auto& line : lines) {
+            line.rect.setY(yOffset);
+            yOffset += line.rect.height() + attachmentItemMargin;
+        }
+    }
+
+    yOffset -= attachmentItemMargin;
+
+    contentYOrigin = (attachmentRect.height() / 2) - (yOffset / 2);
+}
+
+#endif // PLATFORM(IOS_FAMILY)
+
+void AttachmentLayout::addLine(CTFontRef font, CTLineRef line, bool isSubtitle = false)
+{
+    CGRect lineBounds = CTLineGetBoundsWithOptions(line, excludeTypographicLeading ? kCTLineBoundsExcludeTypographicLeading : 0);
+    CGFloat trailingWhitespaceWidth = CTLineGetTrailingWhitespaceWidth(line);
+    CGFloat lineWidthIgnoringTrailingWhitespace = lineBounds.size.width - trailingWhitespaceWidth;
+    CGFloat lineHeight = lineBounds.size.height + (excludeTypographicLeading ? lineBounds.origin.y : 0);
+    lineHeight = isSubtitle ? lineHeight : CGCeiling(lineHeight);
+    
+    CGFloat xOffset = (widthPadding / 2) - (lineWidthIgnoringTrailingWhitespace / 2);
+    LabelLine labelLine;
+    labelLine.font = font;
+    labelLine.line = line;
+    labelLine.rect = FloatRect(xOffset, 0, lineWidthIgnoringTrailingWhitespace, lineHeight);
+    
+    lines.append(labelLine);
+}
+
+void AttachmentLayout::buildWrappedLines(String& text, CTFontRef font, NSDictionary *textAttributes, unsigned maximumLineCount)
+{
+    if (text.isEmpty())
+        return;
+    
+    RetainPtr<NSAttributedString> attributedText = adoptNS([[NSAttributedString alloc] initWithString:text attributes:textAttributes]);
```

Comment:
Could use lots more `auto` in this function.

Reply:
Fixed.